STEP4

satsuki2200 opened the pull request:
## What
<!--- Write the change being made with this pull request --->
- 4-2
  - Add `category` to `POST /item`
- 4-3
  - Implement `GET /items`
- 4-4
  - Add `image` category to `GET /items` and `POST /items`
  - Add image storage process.
- 4-5
  - Implement `GET /items/<item_id>`

## Prove

<img width="934" alt="image" src="https://github.com/user-attachments/assets/1916225a-2dfb-40db-bab4-0263324f614e" />

## CHECKS :warning:

Please make sure you are aware of the following.

- [x] **The changes in this PR doesn't have private information

satsuki2200 committed:
4-2: add category and save data to json

satsuki2200 committed:
4-3: feat GET /items

satsuki2200 committed:
4-4: add image

satsuki2200 committed:
4-5: GET /items/<item_id>

YadaYuki commented on go/app/infra.go:
```diff
@@ -12,6 +16,12 @@ var errImageNotFound = errors.New("image not found")
 	ID   int    `db:"id" json:"-"`
 	Name string `db:"name" json:"name"`
+	Category string `db:"category" json:"category"`
+	Image_Name string `db:"image_name" json:"image_name"`
```
When writing go ,you should name in mixed cap(camel case)!: https://google.github.io/styleguide/go/guide#mixedcaps
```suggestion
	ImageName string `db:"image_name" json:"image_name"`
```

YadaYuki commented on go/app/server.go:
```diff
@@ -180,6 +279,55 @@ func parseGetImageRequest(r *http.Request) (*GetImageRequest, error) {
+	Image_Name string `json:"image_name"`
+}
+
+func (s *Handlers) GetItemToId(w http.ResponseWriter, r *http.Request) {
```
GetItem**By**ID is better naming , i think!
```suggestion
func (s *Handlers) GetItemById(w http.ResponseWriter, r *http.Request) {
```

YadaYuki commented on go/app/server.go:
```diff
@@ -180,6 +279,55 @@ func parseGetImageRequest(r *http.Request) (*GetImageRequest, error) {
+		http.Error(w, "invalid request", http.StatusBadRequest)
+		return
+	}
+	itemId, _ := strconv.Atoi(uri[2])
```
Could you add error handing here, too ?

YadaYuki commented on go/app/server.go:
```diff
@@ -180,6 +279,55 @@ func parseGetImageRequest(r *http.Request) (*GetImageRequest, error) {
+	}
+
+	// find item by itemId
+	resp := GetItemToIdResponse{
+		Name: items[itemId].Name,
+		Category: items[itemId].Category,
+		Image_Name: items[itemId].Image_Name,
+	}
```
Please check the length of `items` is less than `itemId` before reference `items[itemId]` like:
```go
if len(items) < itemId {
    ... throw invalid item id error ...
}
```
and `itemId - 1` may be correct ?
```go
resp := GetItemToIdResponse{
    Name: items[itemId - 1].Name,
    Category: items[itemId - 1].Category,
    Image_Name: items[itemId - 1].Image_Name,
}
```

satsuki2200 committed:
fix variable names

satsuki2200 committed:
fix: add error handling

satsuki2200 committed:
fix itemId validation logic

pandakun26 commented on go/app/server.go:
```diff
@@ -159,7 +224,41 @@ func (s *Handlers) storeImage(image []byte) (filePath string, err error) {
+	if err != nil {
+		slog.Error("failed to write image: ", "error", err)
+		return
+	}
```
Could you please move this storing logic to `StoreImage` function in infra.go?

pandakun26 commented on go/app/server.go:
```diff
@@ -180,6 +279,66 @@ func parseGetImageRequest(r *http.Request) (*GetImageRequest, error) {
+	if len(uri) < 3 {
+		http.Error(w, "invalid request", http.StatusBadRequest)
+		return
+	}
```
hmm I think there are much better ways to extract the path value `id` here.

since you are using mux, I think you can directly get the id like this

`itemIdStr := r.PathValue("id")`

satsuki2200 replied:
I didn't know how to do that. I've fixed it!

pandakun26 commented on go/app/infra.go:
```diff
@@ -36,8 +47,43 @@ func NewItemRepository() ItemRepository {
+
+	// decode bytes to map
+	var decodeData JsonFormat
+	json.Unmarshal(bytes, &decodeData)
```
I think this same logic is repeated at many places, let's follow DRY principle & extract this logic to a function which will take `fileName` as input and return `[]Item, error`

pandakun26 commented on go/app/server.go:
```diff
@@ -159,7 +224,41 @@ func (s *Handlers) storeImage(image []byte) (filePath string, err error) {
+	hashString := fmt.Sprintf("%x", hashSum)
+
+	// build image file path
+	filePath, buildImagePathError := s.buildImagePath(hashString + ".jpg")
```
could you please extract `hashstring + ".jpg"` to a separate variable like `fileName` and then replace all usage of `hashString + ".jpg"` with `fileName`

That would make it much easier to read and more maintainable

pandakun26 reviewed (commented):
Nicely done! Left some comments.

pandakun26 commented on go/app/server.go:
```diff
@@ -76,9 +81,46 @@ func (s *Handlers) Hello(w http.ResponseWriter, r *http.Request) {
+	bytes, _ := io.ReadAll(jsonFile)
+
+	// decode bytes to map
+	var decodeData map[string][]Item
+	json.Unmarshal(bytes, &decodeData)
+
+	var items []Item
+	for _, v := range decodeData {
+		items = append(items, v...)
+	}
```
sorry why do we need `map[string][]Item`, cant we just simply do this

```	var items []Item
	if err := json.Unmarshal(bytes, &items); err != nil {
		return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
	}
```

let me know if im missing anything 🙇

pandakun26 replied:
sorry @satsuki2200 , on the second look I understand now why you were using `map[string][]Item`, so you can ignore the above comment.
Sorry for the confusion

satsuki2200 committed:
fix store image logic

satsuki2200 committed:
refactor GetItemById handler to use mux for extracting itemID

pandakun26 reviewed (approved):
Great job, LGTM 🚀 
there's only [one comment left ](https://github.com/satsuki2200/mercari-build-training/pull/3#discussion_r1981804659)to be addressed, so once you address it, you can proceed to merge this PR

satsuki2200 committed:
refactor: extract repeated logic into a function

satsuki2200 commented:
> Great job, LGTM 🚀
> there's only [one comment left ](https://github.com/satsuki2200/mercari-build-training/pull/3#discussion_r1981804659)to be addressed, so once you address it, you can proceed to merge this PR

Addressed the remaining comment. Merging now, thanks!
